### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by Rifdhan**

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+const MousePresence : any = {
+    /**
+     * Add a callback for the mouse disappear event
+     * @Param {() => void} cb - The callback to be added
+     * @Return {() => void} A reference to the same callback
+     */
+    onDisappear: (cb: () => void): () => void => {
+        onDisappearCallbacks.push(cb);
+        return cb;
```
Comment (ashubham) [inline index.ts]: Should we follow the pattern where callback queuer returns a deregisterer ?

File: index.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
+"use strict";
+
+let onDisappearCallbacks: (() => void)[] = [];
```
Comment (ashubham) [inline index.ts:3]: These could be WeakSets instead of arrays. That makes sure the callbacks do not leak memory.

File: index.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        onDisappearCallbacks = onDisappearCallbacks.filter(cb => cb !== target);
+        onAppearCallbacks = onAppearCallbacks.filter(cb => cb !== target);
+    },
+
+    /**
+     * Dispatch a mouse disappear event and add a listener to dispatch a mouse appear event
+     * @Return {void}
+     */
+    dispatchDisappear: (): void => {
+        if(cursorIsVisible) {
```
Comment (ashubham) [inline index.ts]: nit: I generally prefer a shortcut for readability.
```
if(!cursorVisible) {
  return;
}

// Main logic goes here
```

File: index.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                }
+            }, { once: true });
+        }
+    },
+
+    /**
+     * Check if the mouse is present at time of calling
+     * @Return {boolean} Whether the mouse is present
+     */
+    isPresentNow: (): boolean => {
```
Comment (ashubham) [inline index.ts]: `Now` seems to be redundant .. just `isPresent` should be good enough IMO.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,99 @@
...
+Determine if the mouse cursor is present or not (according to the last reported
+appearance/disappearance event).
+
+
+# Examples
+
+Component A:
+
+```javascript
+let MousePresence = require("mouse-presence");
```
Comment (ashubham) [inline README.md]: nit: maybe the example could use `import` syntax. The `require` syntax is non standard.

**Review by ashubham — COMMENTED** [review]
I had an idea regarding the contract here, what if we make this do more and accept a HTMLElement as a param and then sets up event listeners on appropriate events on that. (for eg. `keydown` for Input/Text).

That makes this module usable out of the box. Though, I agree might be less optimal for us as there would be two event listeners.

File: index.ts (hunk 1 above)
Comment (Rifdhan) [reply index.ts]: Sounds good, done

File: index.ts (hunk 3 above)
Comment (Rifdhan) [reply index.ts]: Done

File: index.ts (hunk 4 above)
Comment (Rifdhan) [reply index.ts]: Done

File: README.md (hunk 5 above)
Comment (Rifdhan) [reply README.md]: Good idea, done

File: index.ts (hunk 2 above)
Comment (Rifdhan) [reply index.ts:3]: As discussed offline, WeakSets are not iterable so cannot be used here

**Comment by Rifdhan** [comment]
Added a convenience function to do what you described, as per our discussion. Preserved the existing functionality as well, to support a more generic use case.

File: index.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            cursorIsVisible = true;
+            onAppearCallbacks.forEach(cb => cb());
+        }, { once: true });
+    },
+
+    /**
+     * Check if the mouse is present at time of calling
+     * @Return {boolean} Whether the mouse is present
+     */
+    isPresent: (): boolean => {
```
Comment (surgicaI) [inline index.ts]: nit: can write in shortcut
isPresent: (): boolean => cursorIsVisible,

File: index.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    },
+
+    /**
+     * Add a keydown listener to the provided HTMLElement, which will trigger a dispatchDisappear
+     * call when fired.
+     * @Param {HTMLElement} el - The element for which the keydown listener should be added to
+     * @Return {() => void} A deregisterer for the listener that will be attached
+     */
+    dispatchDisappearOnKeydown: (el: HTMLElement): () => void => {
+        let listener = () => { MousePresence.dispatchDisappear() };
```
Comment (surgicaI) [inline index.ts]: nit: let listener = (): void =>

File: index.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        el.addEventListener('keydown', listener);
+        return () => { el.removeEventListener('keydown', listener) };
```
Comment (surgicaI) [inline index.ts]: (): void

**Review by surgicaI — APPROVED** [review]
looks good to me

File: index.ts (hunk 6 above)
Comment (Rifdhan) [reply index.ts]: Done

File: index.ts (hunk 7 above)
Comment (Rifdhan) [reply index.ts]: Done

File: index.ts (hunk 8 above)
Comment (Rifdhan) [reply index.ts]: Done

**Pull request merged**